### ISSUE TITLE
How to cross compile to different architectures?

### ISSUE THREAD

**Issue by Jonathas-Conceicao** [Arc-blroth/gradle-cargo-wrapper#2]
I have setup the plugin here to build a java lib using [flapigen](https://github.com/Dushistov/flapigen-rs) and it works perfectly when I'm building for the native environment. One of the subprojects setups a rust lib and uses `cargo-wrapper` to build the crate, while another subproject that depends on the rust one builds the java lib adding the rust generated library to it's jar output.

I am now trying to setup my project's CI to also cross compile to other architectures, e.g., build to 32-bit Linux from a 64-bit container, but I don't really know how to properly set this up on the gralde config. I've tested the `attributes` property and it does work for passing `--target <triple>`to the Cargo command, but having a fixed value there is not ideal for setting up a CI that generate builds for a few different architectures.

I've tried setting something up on the `outputs` variable, but I'm not sure that this is the right place for my use case. I would expect there to be a way to specify on the command line the desired target for the build, or maybe define a new task that would build the project with its own version of the `attributes` field set, but I haven't found a way if doing either.

I'm not that experienced with gradle, so I might be missing something obvious.

**Comment by Arc-blroth** [comment]
Gradle's buildscripts are actually scripts written in the [Groovy](https://groovy-lang.org/) programming language (not sure if you already knew that), so you can use all of the features of Java and Groovy within a buildscript.

If you wanted to set up something where you could specify a specific Rust target for a build, you could write
```gradle
cargo {
    def targetTriple = System.getProperty("CARGO_WRAPPER_RUST_TARGET")
    if(targetTriple == null) {
        throw new GradleException("Target triple not specified!")
        // or you could also set the targetTriple to some default
        // value rather than aborting here
    }

    def fileExtension
    switch(targetTriple.split('-')[2]) {
        case 'windows':
            fileExtension = 'dll'
            break
        case 'darwin':
            fileExtension = 'dylib'
            break
        default:
            fileExtension = 'so'
            break
    }

    crate = projectDir.path
    
    // (assuming that you typo'd "attributes" and meant "arguments" here)
    arguments = ['--target', targetTriple]

    // build results are expected to be in target/<triple>/<profile>/
    // when passing `--target`, so this map must specify targetTriple
    // as a key to allow cargo-wrapper to find the output of `cargo build`
    outputs = [(targetTriple): 'wrapper_example.' + fileExtension]
}
```
which would let you run
```sh
./gradlew -DCARGO_WRAPPER_RUST_TARGET=<target triple> build
```
on the command line in your CI to build with a particular target triple.

---

On the other hand, if you wanted to set up building multiple architectures of the same library into a single jar, then you would need to register multiple `CargoTask`s manually, which isn't currently possible since the `CargoTask` constructor is protected. If this is what you were looking for, feel free to ask me to change CargoTask's visibility and update this plugin.

**Comment by Jonathas-Conceicao** [comment]
This script you've provided fits very well my needs atm, that's better than having predefined tasks. Thanks for the help!

**Comment by bluebear94** [comment]
> If this is what you were looking for, feel free to ask me to change CargoTask's visibility and update this plugin.

Unfortunately, I need to do this.

**Comment by Arc-blroth** [comment]
@bluebear94 I published version 1.1.0 of the plugin, which makes `CargoTask#configure` public.

**Comment by RXminuS** [comment]
@Arc-blroth I feel very daft, but I can't quite get it right...how exactly would I configure multiple build tasks so I can cross-compile to multiple targets?
